### ISSUE TITLE
Data Set Preprocessing Issues

### ISSUE THREAD

**Issue by cjc2412** [XingfuWang/CSNet#1]
(I am a first-year graduate student currently working on reproducing your CSNet paper, and I've encountered some issues. I'd like to seek your advice.  

I really like your CSNet. I have already trained it on the BCI Competition IV-2a dataset, and now I hope to apply it to EEG emotion recognition and ERN detection. However, my coding skills are limited, and I'm struggling with how to process the data properly.  

I mainly have three questions to clarify:  

1. How exactly is the SEED dataset divided? The paper mentions 5-fold cross-validation. Is it performed within-subject (each subject individually with 5-fold) or across-subjects (all subjects mixed together and then divided into 5 folds)?  

2. How should the three sessions be handled? Should the 15 subjects × 3 sessions = 45 files be merged into one large array before division, or should the three sessions be merged within each subject first, and then divided?  

3. Are there any special processing steps for the ERN data? I read in the paper that only baseline correction is applied?  

Also, if it's convenient, could you share the code for data processing and training? I promise it will only be used for learning purposes and will not be distributed. Of course, if you don't have time, that's completely fine—even some guidance on the questions above would be greatly appreciated!  

I apologize for bothering you, as I know you are very busy. Thank you so much regardless!  

Best regards!

**Comment by XingfuWang** [comment]
Hello, I'm very glad that you like my algorithm. Below I will answer your questions step by step:


1. Each subject's dataset in the SEED dataset is divided into five folds for cross-validation. That is, the results on the SEED dataset in the original paper are with-subject, not cross-subject. I have released the preprocessing of the SEED dataset in this code repository, named `load_seed.py`.


2. We merged the data from the three sessions and then performed 5-fold cross-validation. Specifically, we used `load_seed.py` to load the .mat format data and save it as .npz format data. Then, in `main`, we loaded the .npz format data and used `StratifiedKFold(n_splits=5)` from the sklearn library to split the dataset, and took the average of the 5 folds as the result.


3. The ERN signals only underwent simple baseline correction. It should be noted that the ERN task uses the AUC metric instead of Accuracy.


I'm very happy that you like my work. If you have any other questions, I'll be happy to answer them. Wish you success in your studies!

**Comment by cjc2412** [comment]
Thank you so much for your kind and detailed previous response! Your guidance has clarified many of my confusions and given me great encouragement. I truly appreciate your patience and support.
Following your advice, I have successfully processed the SEED dataset using load_seed.py and am currently working on within-subject 5-fold cross-validation. The data structure is now much clearer to me.
I still have a few more questions and would be extremely grateful if you could spare some time to help:
1. BCI-ERN Dataset Processing Code
If it's not too much trouble, could you please share the processing code for the ERN dataset? I recall you mentioned it only requires baseline correction, but my coding skills are quite limited (I can read and understand code, and modify existing code, but writing from scratch or even with AI assistance is quite difficult for me). Having your reference code would be an invaluable learning resource for me.
2. Reproducing Comparative Experiments
Regarding the baseline methods in your paper, if their original papers typically don't provide complete code. When you conducted these comparisons, did you:
Re-implement them from scratch based on the paper descriptions?
Use third-party implementations (e.g., from GitHub)?
Or have other recommended methods for fair comparison?
Or ask the original authors for their code?
3. Cholesky Decomposition Implementation
Regarding the Cholesky decomposition in your paper, do you directly use the official PyTorch function L = torch.linalg.cholesky(Xb)? Or do you define custom matrix algorithms for the decomposition? I noticed your class CholeskyOperations: in your code. Due to my computer's relatively poor performance, using the official Cholesky decomposition is quite slow and sluggish. I would like to know whether you used custom-defined matrix algorithms to accelerate the Cholesky decomposition?
I have learned a great deal from your work. Your work is truly excellent, I like it very much, and I have benefited greatly. I was very touched by your blessing, which made me feel very warm. Thank you again for your time and generosity!
Best regards,

**Comment by XingfuWang** [comment]
Hello! I'm delighted to see your passion for scientific research. Below I will answer your questions step by step:
1, I have provided the baseline correction code for BCI_ERN, named detrend.py, along with a simple example. You can download it now.
2, For the baseline comparisons in the paper, we used the authors' provided code or results whenever possible. Many of these models are available in the Braindecode library — for instance, you can directly obtain ShallowNet by calling from braindecode.models import ShallowFBCSPNet, and reproduce the results by replacing your own model with the target model in main. Additionally, for models not provided in Braindecode (such as CRGNet, TensorCSPNet, and GraphCSPNet), we invested significant effort in reproducing the authors' open-source code.
3, We used official PyTorch functions. To ensure good decoding performance, we set relatively large initial model parameters. You can reduce some parameters (e.g., the spatial_merge parameter) — if you specify spatial_merge=16, the model will still yield excellent results, and the Cholesky decomposition will run faster.
If you have any further questions, please feel free to ask. Wish you all the best!

**Comment by cjc2412** [comment]
Thank you once again for your detailed and prompt previous response. It has been incredibly helpful.

I am currently attempting to reproduce the results for EEG_Conformer on the BCI Competition IV 2a dataset using the models provided in Braindecode. I have a specific question regarding the experimental setup for the baseline comparisons in your paper that I would like to ask you.

My data processing pipeline is identical to the one used for the CSNet in Braindecode (using the BCI IV 2a dataset, segmented into 4-second intervals, with no preprocessing except converting the unit from V to µV). I also used the same hyperparameters (learning rate=0.0005, epochs=100, batch size=64). However, the accuracy I achieved for subject 1 was only 67.01%, which is significantly lower than the 88.19% reported in your paper for the EEG_Conformer model under a unified setting. Furthermore, my average accuracy was only 64.15%, also much lower than the average accuracy of 78.66% for EEG_Conformer reported in your Table 5.

When you mentioned that you "used the authors' provided code or results whenever possible" for baseline comparisons, could you please clarify what this means in practice? Does it mean that the data processing and splits were kept exactly the same across all models in your setup, while batch size and learning rate were referenced from the original papers? Or, was the data processing also referenced and adjusted according to the original papers of the respective models?

I ask because upon reviewing the original EEG_Conformer paper, I found that its reported accuracy is consistent with the results in your Table 5. The same holds true for IFNet and LMDANet – the accuracies match perfectly. However, their data processing, learning rates, and training schedules differ from the setup described for your dataset. Does this mean that when comparing SOTA models in the paper, you followed the practice of prioritizing the models' originally published optimal hyperparameters, or fine-tuning them for optimal performance on the current dataset, rather than retraining all models under strictly uniform experimental conditions? Generally, isn't the standard practice to retrain and reproduce all models under identical conditions?

I haven't yet compared the models for the SEED and ERN datasets. Was the accuracy in the tables handled similarly for those? Please forgive me if my question seems forward. I am genuinely unfamiliar with the nuances of academic writing and sincerely hope to learn more from your expertise. Thank you for your time and valuable insights.

Sincerely,

**Comment by XingfuWang** [comment]
Hello! I'm glad to see your reply. Below are answers to your questions and concerns:
1. The performance of the EEGConformer model is indeed highly volatile. We encountered similar issues during our reproduction process — in the original paper, the accuracy of EEGConformer fluctuates sharply during training. Therefore, it is difficult to determine whether this is due to reproduction challenges or inherent flaws of the model itself. We thus had to adopt the original results provided by the authors for comparison, to better highlight the competitiveness of our model.
2. Our statement "used the authors' provided code or results whenever possible" is accurate. Ideally, we strived to use identical data processing pipelines, dataset splitting methods, and hyperparameters. However, when conditions did not allow (e.g., some models rely on frequency-domain features, making it impossible to apply the same preprocessing as time-domain EEG-based models), we ensured the most fair comparison possible. When unavoidable discrepancies arose, we made reasonable concessions — but we prioritized selecting the higher value between the authors' reported results and our reproduced results. This approach helps highlight our model’s competitiveness relative to others while minimizing potential disputes.
3. Your observation is precise. As you mentioned with IFNet, enforcing a unified preprocessing pipeline would severely contradict the authors’ multi-band design philosophy, constituting an unfair comparison. For this reason, we opted to use the authors’ reported results.
4. We applied similar handling to the SEED and ERN datasets to ensure the fairest possible comparisons.
Wish you a pleasant learning experience! Feel free to ask if you have any further questions.

**Comment by cjc2412** [comment]
First and foremost, I would like to express my sincere gratitude for your previous assistance. Your guidance has been immensely helpful to me. Due to differences in hyperparameter selection and data processing, it is indeed challenging to achieve the accuracy reported in other papers. Regarding the comparative experiments in my paper, I would like to ask: Should I conduct experiments under the same hardware environment and then compare the results with those reported in others' papers? To highlight competitiveness, is it acceptable to select the better result between the two? If reviewers inquire about the accuracy comparison, would simply submitting the saved model weights be sufficient?

Your paper is exceptionally well-written, and I truly admire its fluency. As a researcher who has built a model but struggles with writing, I am eager to learn from your writing methods. Additionally, while using Braindecode to run EEGConformer on the BCI IV 2a dataset, my accuracy consistently remains below 20%, which has been quite troubling. Below is my code—I am unsure where the issue lies:

python
[import scipy.io
import mne
import numpy as np

root_dir = r"D:\EEG_data\BCICIV_2a_gdf\A01T.gdf"
raw = mne.io.read_raw_gdf(root_dir, preload=True)
eog_channels = ['EOG-left', 'EOG-central', 'EOG-right']
raw.drop_channels(eog_channels)
label_path = r"D:\EEG_data\BCICIV_2a_lables\true_labels\A01T.mat"
mat_data = scipy.io.loadmat(label_path)
labels = mat_data['classlabel'].flatten() - 1
events, event_id = mne.events_from_annotations(raw)
extract_events = events[events[:, 2]  == event_id['768']]
sfreq = raw.info['sfreq']
cue_delay = 2.0
tmin = cue_delay
tmax = cue_delay + 4.0 - 1/sfreq 
n_samples = int((tmax - tmin) * sfreq)
epochs = mne.Epochs(raw,events=events,event_id={'768': event_id['768']},  
tmin=tmin,tmax=tmax,baseline=None,preload=True,
)
X = epochs.get_data() * 1e6  
y = labels[:len(X)]  
from skorch.dataset import ValidSplit
from braindecode import EEGClassifier
net = EEGClassifier(
    "EEGConformer",train_split=ValidSplit(0.2),max_epochs=20,
)
net.fit(X, y)]
Furthermore, I would like to seek your advice on the following: How can I systematically learn academic writing? How should I design comparative and ablation experiments? How can I effectively highlight the value of my paper to improve its chances of acceptance by journals?

I apologize for asking so many questions at once. Lately, I have been feeling quite lost  in my research. Thank you once again for your continuous guidance and encouragement. I look forward to learning more from you.

**Comment by XingfuWang** [comment]
Hello, you're welcome. Regarding your questions:
1. There's no need to conduct experiments in an identical hardware environment, as this is hardly achievable and reviewers won't impose such strict requirements. Try to select the best possible comparison results—this not only respects the algorithm proposers but also demonstrates the competitiveness of your own model. As long as the results are objective and authentic, reviewers will recognize them.
2. I'm not very familiar with your code net = EEGClassifier("EEGConformer", train_split=ValidSplit(0.2), max_epochs=20), but max_epochs=20 is far from sufficient for EEGConformer. You need at least 300 epochs.
3. Generally speaking, academic writing is included in graduate courses. If not, you can search online to learn. After completing your writing, you can use generative models to check if your writing is rigorous and identify areas for improvement, gradually enhancing your professionalism. Ablation experiments involve designing one or more key modules, proving that they are indeed effective, and showing that the model's performance degrades when they are removed. To increase the chances of your paper being accepted: first, download high-quality articles from relevant journals to imitate their writing style; second, highlight the improvements of your research compared to previous studies; finally, conduct thorough experimental verification.
Wish you a pleasant learning experience!

**Comment by cjc2412** [comment]
Thank you once again for your previous thoughtful and detailed response—it was truly helpful. I’m reaching out again because I’ve encountered some new confusion while designing comparative experiments, and I would sincerely appreciate your guidance.

1. Regarding the design of comparative experiments:
I’m unsure which of the following approaches is more appropriate:

Training my model and the comparison models on my own preprocessed dataset (with consistent hyperparameters such as learning rate, number of epochs, etc.), and presenting the results in a table;

Training both my model and the comparison models on my own processed dataset, while also referencing the experimental results reported in the original paper, then taking the higher results between the comparison models and the original paper for a fair comparison;

Or, conducting a three-way comparison: training both my model and the comparison models on my own processed dataset, preprocessing the data according to the comparison methods' own preprocessing protocols before experimentation, and also including the experimental results from the original paper, then selecting the highest among these three. The third approach seems quite time-consuming to me.

I want to ensure my comparisons are both rigorous and respectful of prior work, so I would be grateful for your advice on which approach is considered more standard or convincing in peer-reviewed research.

2. On selecting models and papers for comparison:
I also find it challenging to choose suitable baseline models and papers for comparison. For example, for geometric learning-based methods like CSNet, should I compare it mainly with other geometric learning models (e.g., SPDNet, MAtt), or should I also include Euclidean space-based models? How do you typically decide on the scope and variety of baselines to ensure a comprehensive yet focused comparison?

3. Regarding SSVEP processing:
Lastly, if you are familiar with SSVEP (Steady-State Visual Evoked Potential) processing, I would appreciate any recommendations on methods, tools, or recent papers that you consider effective or noteworthy. If this falls outside your expertise, no worries—I completely understand.

Thank you for your time and patience. Your experience and suggestions mean a lot to me as I navigate this research journey.

Warm regards,

**Comment by XingfuWang** [comment]
I am glad to receive your reply, and below is the response to your questions:
1. When comparing experimental results, if the comparison models can adopt the same preprocessing and hyperparameters as the model we proposed, directly using the identical processing and hyperparameters can be recognized, which also saves us time and effort. However, as I mentioned earlier, many models are not compatible with the preprocessing we use, so it is inevitable to invest additional effort. For example, I have devoted some time and energy to reproducing TensorCSPNet and GraphCSPNet in accordance with their original papers.
2. Except for Riemannian geometry-based methods, the vast majority of models can be regarded as Euclidean geometry-based methods. This is because all calculations of these models are performed based on Euclidean space, without specifically considering their non-Euclidean characteristics.
3. I am not very familiar with SSVEP. However, to my knowledge, Xiaorong Gao (Tsinghua University) and Yijun Wang (Chinese Academy of Sciences) have conducted outstanding research in this field, and you may refer to their publications for more insights.
Wish you a pleasant study!

**Comment by cjc2412** [comment]
Thank you very much for your thoughtful and detailed answer! Your guidance has given me a clearer understanding of how to design comparative experiments.

Based on your explanation, I now further understand that “identical parameters” ≠ fairness—in fact, this is often a common source of unfairness. If we aim for a fair comparison, the core principle should be to allow each model to achieve its best possible performance or theoretical upper limit. Therefore, a more reasonable fairness constraint would be: under the same data split, evaluation metrics, optimizer type (e.g., both using Adam), and hardware/software environment (such as the same GPU and PyTorch/CUDA version), each model should be allowed to use its most suitable learning rate, number of training epochs, batch size, loss function (some may include center loss), and learning rate scheduler (e.g., cyclical or cosine annealing), and so on. I wonder if my understanding is correct?

Additionally, I would like to ask: how should comparison models be appropriately selected? My current idea is to look for papers in the target journal that use the same dataset, and then refer to the set of models they compare against, incorporating them into my comparative framework. Do you think this is a reliable strategy?

Finally, thank you again for recommending scholars in the field of SSVEP. I will carefully study the relevant papers by Professor Xiaorong Gao from Tsinghua University, Professor Yijun Wang from the Chinese Academy of Sciences, and other research teams.

I truly appreciate the time you have taken and the valuable experience you have shared. I look forward to your further guidance.

Wishing you all the best in your work!

**Comment by XingfuWang** [comment]
I’m glad to receive your reply. Below are the answers to your questions:

1. I think your understanding is correct.
2. When selecting a model, you should try to choose the one that is most relevant to your research topic, preferably an open-source model. You can also refer to the comparative models used in related papers, which is an excellent source of reference.

Wish you all the best!
